### PR TITLE
PLANNER-826: Guided Rule Editor doesn't have reserved scoreHolder variable

### PR DESCRIPTION
**Pull request by mcimbora**
@manstis This simply displays a warning message when user tries to use 'scoreHolder' (or other) reserved variable name in GRE/GDT. Individual domains (e.g. planner) need to provide PatternBindingValidator implementations. Would you mind taking a look? Thanks in advance.

### CODE REVIEW

**Comment by mcimbora** [comment]
Part of:
- https://github.com/kiegroup/drools-wb/pull/639
- https://github.com/kiegroup/optaplanner-wb/pull/224

**Comment by jomarko** [comment]
@mcimbora I will have a look also, if you don't mind.

**Comment by mcimbora** [comment]
@jomarko Go ahead, thanks.

**Comment by jomarko** [comment]
@mcimbora Thanks for this PR. It works perfect for guided rule editor. However within guided decision tables, the situation is a little bit complicated. In guided tables user can define condition/action column also without usage of guided rule editor. It means that you PR don't cover two places:
- PatternPage, I think change will be needed [here](https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/PatternPage.java) or [here](https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/modals/NewPatternPresenter.java) 
- ValueOptionsPage, the field binding, that binds field of the pattern to some name, I think change will be needed [here](https://github.com/kiegroup/drools-wb/blob/master/drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/column/pages/ValueOptionsPage.java)
screenshot from 2017-10-05 09-23-53

screenshot from 2017-10-05 09-18-53

**Comment by manstis** [comment]
@mcimbora @jomarko I've not looked at this PR yet, but based on @jomarko comment I suggest you take a look at ```BRLRuleModel``` and the existing use of it. It adapts a ```GuidedDecisionTable52``` into a ```RuleModel``` and checks for bound variables etc span both models.

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/RuleModeller.java [hunk 1]
```diff
@@ -118,6 +122,7 @@ public RuleModeller(final RuleModel model,
                         final boolean isReadOnly) {
         this(model,
              Collections.emptyList(),
+             Collections.emptyList(),
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/RuleModeller.java]: You should check the operation of "Guided Rule Templates" - as without the validators being passed here - you could probably use an OptaPlanner reserved binding in "Guided Rule Templates"... I guess you don't care for DRL files re-using OptaPlanner bindings?

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/factPattern/PopupCreator.java [hunk 2]
```diff
@@ -106,71 +110,77 @@ public boolean isBindable() {
...
+        ok.addClickHandler(new ClickHandler() {
+            public void onClick(ClickEvent event) {
                 String var = varName.getText();
-                if ( modeller.isVariableNameUsed( var ) ) {
-                    Window.alert( GuidedRuleEditorResources.CONSTANTS.TheVariableName0IsAlreadyTaken( var ) );
+                if (modeller.isVariableNameUsed(var)) {
+                    Window.alert(GuidedRuleEditorResources.CONSTANTS.TheVariableName0IsAlreadyTaken(var));
+                    return;
+                }
+                String validationTextToDisplay = PatternBindingValidatorUtils.getValidationTextToDisplay(getModeller().getPatternBindingValidators(),
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/factPattern/PopupCreator.java]: I'd like to have the check for used variables in one place.. so I'd like to see use of the new validators moved to ```modeller.isVariableNameUsed(var)``` as on L#139 (and necessary changes to return the error message). Probably more refactoring than tagging it on here. Ideally it'd be best to move the whole "duplicate variable name" implementation to use the new ```PatternBindingValidator``` behaviour with core implementations.

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/factPattern/PopupCreator.java [hunk 3]
```diff
@@ -418,37 +427,43 @@ private String getFactType( FactPattern fp,
...
+            bindVar.addClickHandler(new ClickHandler() {
+                public void onClick(ClickEvent event) {
                     String var = varTxt.getText();
-                    if ( modeller.isVariableNameUsed( var ) ) {
-                        Window.alert( GuidedRuleEditorResources.CONSTANTS.TheVariableName0IsAlreadyTaken( var ) );
+                    if (modeller.isVariableNameUsed(var)) {
+                        Window.alert(GuidedRuleEditorResources.CONSTANTS.TheVariableName0IsAlreadyTaken(var));
+                        return;
+                    }
+                    String validationTextToDisplay = PatternBindingValidatorUtils.getValidationTextToDisplay(getModeller().getPatternBindingValidators(),
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/factPattern/PopupCreator.java]: Likewise. re: my comments above.

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/validator/PatternBindingValidator.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.drools.workbench.screens.guided.rule.client.editor.validator;
+
+import java.util.Collection;
+
+import org.guvnor.common.services.shared.validation.model.ValidationMessage;
+
+public interface PatternBindingValidator {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/validator/PatternBindingValidator.java]: You also use implementations of this ```interface``` for **field** binding validation; so perhaps the name could be improved.

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/validator/PatternBindingValidator.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package org.drools.workbench.screens.guided.rule.client.editor.validator;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/validator/PatternBindingValidator.java]: License

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/validator/PatternBindingValidatorUtils.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package org.drools.workbench.screens.guided.rule.client.editor.validator;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/editor/validator/PatternBindingValidatorUtils.java]: License

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/widget/ActionInsertFactWidget.java [hunk 7]
```diff
@@ -277,6 +278,12 @@ public void onClick(ClickEvent event) {
                     Window.alert(GuidedRuleEditorResources.CONSTANTS.TheVariableName0IsAlreadyTaken(var));
                     return;
                 }
+                String validationTextToDisplay = PatternBindingValidatorUtils.getValidationTextToDisplay(getModeller().getPatternBindingValidators(),
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/widget/ActionInsertFactWidget.java]: This too can be moved to ```RuleModeller```; as called [here](https://github.com/mcimbora/drools-wb/blob/e1b0183d6b588e3fd0c79da9bac0d20a43979a7f/drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/widget/ActionInsertFactWidget.java#L277).

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/widget/ExpressionBuilder.java [hunk 8]
```diff
@@ -499,6 +500,12 @@ public void onClick( ClickEvent event ) {
                                 Window.alert( GuidedRuleEditorResources.CONSTANTS.TheVariableName0IsAlreadyTaken( var ) );
                                 return;
                             }
+                            String validationTextToDisplay = PatternBindingValidatorUtils.getValidationTextToDisplay(getModeller().getPatternBindingValidators(),
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/widget/ExpressionBuilder.java]: And again [here](https://github.com/mcimbora/drools-wb/blob/e1b0183d6b588e3fd0c79da9bac0d20a43979a7f/drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/main/java/org/drools/workbench/screens/guided/rule/client/widget/ExpressionBuilder.java#L499).

File: drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/test/java/org/drools/workbench/screens/guided/rule/client/validator/PatternBindingValidatorUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package org.drools.workbench.screens.guided.rule.client.validator;
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/test/java/org/drools/workbench/screens/guided/rule/client/validator/PatternBindingValidatorUtilsTest.java]: License

**Comment by ge0ffrey** [comment]
@manstis Do we have someone to handle this PR and the related optaplanner-wb PR - as an exception?
If we wait for Matej's replacement, this PR might be hopelessly stale...

**Comment by manstis** [comment]
@ge0ffrey No we do not have anyone to handle this PR.

**Comment by jomarko** [comment]
@manstis @ge0ffrey if there is now need just for rebasing the PRs I could have a look. If there is still need some deeper development, not sure if manageable from my site.

**Comment by kkufova** [comment]
@manstis, @jomarko, is this still needed? The PR hasn't been touched since 2017. Can we close it?

**Comment by jomarko** [comment]
From my point of view we can close the PR and keep the jira opened.

**Comment by manstis** [comment]
I agree with @jomarko.

**Pull request closed without merging**